bql/grammar/grammar_test.go

```diff
@@ -400,15 +400,131 @@ func TestSemanticStatementGraphClausesLengthCorrectness(t *testing.T) {
+		}
+		if got, want := len(st.ConstructClauses()), entry.want; got != want {
+			t.Errorf("Invalid number of construct clauses for query %q; got %d, want %d; %v", entry.query, got, want, st.ConstructClauses())
+		}
+	}
+}
+
+func TestSemanticStatementReificationClausesLengthCorrectness(t *testing.T) {
+	table := []struct {
+		query     string
+		want_one  int
+		want_two  int
+	}{
+		{
+			query: `construct {?s "predicate_1"@[] ?o1;
+			                      "predicate_2"@[] ?o2.
+				           ?s "predicate_3"@[] ?o3} into ?a from ?b where {?s "old_predicate_1"@[,] ?o1.
+					                                                   ?s "old_predicate_2"@[,] ?o2.
+											   ?s "old_predicate_3"@[,] ?o3};`,
+			want_one:  1,
+			want_two:  0,
+		},
+		{
+			query: `construct {?s "predicate_1"@[] ?o1;
+			                      "predicate_2"@[] ?o2;
+			                      "predicate_3"@[] ?o3.
+				           ?s1 "predicate_1"@[] ?o1;
+				               "predicate_2"@[] ?o2;
+				               "predicate_3"@[] ?o3} into ?a from ?b where {?s "old_predicate_1"@[,] ?o1.
+					                                                    ?s "old_predicate_2"@[,] ?o2.
+											    ?s1 "old_predicate_3"@[,] ?o3};`,
+			want_one:  2,
+			want_two:  2,
+		},
+		{
+			query: `construct {?s "predicate_1"@[2015-07-19T13:12:04.669618843-07:00] ?o1;
+			                      "predicate_2"@[2015-07-19T13:12:04.669618843-07:00] ?o2.
+				           ?s "predicate_3"@[2015-07-19T13:12:04.669618843-07:00] ?o3} into ?a from ?b where {?s "old_predicate_1"@[,] ?o1.
+					                                                                                      ?s "old_predicate_2"@[,] ?o2.
+											                                      ?s "old_predicate_3"@[,] ?o3};`,
+			want_one:  1,
+			want_two:  0,
+		},
+		{
+			query: `construct {?s "predicate_1"@[2015-07-19T13:12:04.669618843-07:00] ?o1;
+			                      "predicate_2"@[2015-07-19T13:12:04.669618843-07:00] ?o2;
+			                      "predicate_3"@[2015-07-19T13:12:04.669618843-07:00] ?o3.
+				           ?s1 "predicate_1"@[2015-07-19T13:12:04.669618843-07:00] ?o1;
+				               "predicate_2"@[2015-07-19T13:12:04.669618843-07:00] ?o2;
+				               "predicate_3"@[2015-07-19T13:12:04.669618843-07:00] ?o3} into ?a from ?b where {?s "old_predicate_1"@[,] ?o1.
+															       ?s "old_predicate_2"@[,] ?o2.
+															       ?s1 "old_predicate_3"@[,] ?o3};`,
+			want_one:  2,
+			want_two:  2,
+		},
+		{
+			query: `construct {?s "predicate_1"@[] ?o1;
+			                      "predicate_2"@[] ?o2;
+			                      "predicate_3"@[?t] ?o3.
+				           ?s1 "predicate_1"@[] ?o1;
+				               "predicate_2"@[] ?o2;
+				               "predicate_3"@[?t] ?o3} into ?a from ?b where {?s "old_predicate_1"@[,] ?o1.
+											      ?s "old_predicate_2"@[,] ?o2.
+											      ?s1 "old_predicate_3"@[?t] ?o3};`,
```

Comment:
?s1 "old_predicate_3"@[?t] ?o3

This is not a valid statement. The valid representation is by using the AT binding keyword. See the example below.

?s1 "old_predicate_3"@[,] AT ?t ?o3